source/opt/licm_pass.cpp

```diff
@@ -0,0 +1,324 @@
+// Copyright (c) 2017 Google Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "licm_pass.h"
+#include "cfg.h"
+#include "module.h"
+
+#include "pass.h"
+
+namespace spvtools {
+namespace opt {
+
+LICMPass::LICMPass(){};
+
+Pass::Status LICMPass::Process(ir::IRContext* context) {
+  bool modified = false;
+
+  if (context != nullptr) {
+    ir_context = context;
+    modified = ProcessIRContext();
+  }
+
+  return modified ? Status::SuccessWithChange : Status::SuccessWithoutChange;
+}
+
+bool LICMPass::ProcessIRContext() {
+  bool modified = false;
+  ir::Module* module = ir_context->module();
+
+  // Process each function in the module
+  for (ir::Function& f : *module) {
+    modified |= ProcessFunction(&f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessFunction(ir::Function* f) {
+  bool modified = false;
+  ir::LoopDescriptor loop_descriptor{f};
+
+  // dom_analysis = ir_context->GetDominatorAnalysis(f, cfg);
+
+  // Process each loop in the function
+  for (ir::Loop& loop : loop_descriptor) {
+    modified |= ProcessLoop(loop, f);
+  }
+  return modified;
+}
+
+bool LICMPass::ProcessLoop(ir::Loop& loop, ir::Function* f) {
+  // Process all nested loops first
+  for (ir::Loop* nested_loop : loop) {
+    ProcessLoop(*nested_loop, f);
+  }
+
+  std::vector<ir::Instruction*> instructions{};
+  GatherAllLoopInstructions(loop, &instructions);
+
+  return ProcessInstructionList(loop, &instructions);
+}
+
+void LICMPass::GatherAllLoopInstructions(
+    ir::Loop& loop, std::vector<ir::Instruction*>* instructions) {
+  for (uint32_t bb_id : loop.GetBlocks()) {
+    ir::BasicBlock* bb = ir_context->get_instr_block(bb_id);
+    for (ir::Instruction& inst : *bb) {
+      if (!DoesOpcodeHaveSideEffects(inst.opcode())) {
```

Comment:
I think it would be better to maintain `instructions` as the set of loop invariant instructions.

at the same time, `instructions` -> `loop_iv_instr` or something like that ?